ipld/resolver/src/provider_cache.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use std::collections::{HashMap, HashSet};
+
+use ipc_sdk::subnet_id::SubnetID;
+use libp2p::PeerId;
+
+use crate::provider_record::{ProviderRecord, Timestamp};
+
+/// Change in the supported subnets of a peer.
+#[derive(Debug, Default)]
+pub struct ProviderDelta {
+    pub added: Vec<SubnetID>,
+    pub removed: Vec<SubnetID>,
+}
+
+impl ProviderDelta {
+    pub fn is_empty(&self) -> bool {
+        self.added.is_empty() && self.removed.is_empty()
+    }
+}
+
+pub struct SubnetProviderCache {
+    /// Maximum number of subnets to track, to protect against DoS attacks, trying to
```

Comment:
I looked at the Gossipsub scoring and man, that's like a nuclear reactor. I reckon it works for over producing peers, peers that send messages that cannot pass basic validation (which is stateless). Problem here is there's no way to tell if agents are peddling complete rubbish, so in the end I just added the _pin_ feature which can be coupled with facts from the ledger. Using that, attacks like this would be pointless I think, so we can just let Gossipsub calm DoS attacks.